Clean up flake8 warnings from doctests

InCogNiTo124 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #19374 

#### Brief description of what is fixed or changed
running `flake8 --doctests sympy/` returns a bunch of errors. This PR is to fix them

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* other
  * Improve code quality by cleaning the doctests up
<!-- END RELEASE NOTES -->

InCogNiTo124 commented:
Most of the unused imports were deleted, but some docstrings were improved, such as `matrices/common.py`, `functions/elementary/trigonometric.py` and some others

also, I only got rid od all F401, I've yet to get rid of other errors (F403, F811, F821)

p.s. sorry for the deleted file, I did it accidentally somehow in terminal but I restored it back immediately

oscarbenjamin commented:
One of the commits deletes the travis fail and another adds it back. If this is merged then that file would be deleted for one commit in the history and would then lose all of its history as view through `git blame`. The commits here need to be squashed so that there is no commit that removes that file.

InCogNiTo124 commented:
I think all errors that are left are those that populate the namespace directly, and frankly I'm not really sure how to approach in handling those other than `#noqa`.

oscarbenjamin commented:
There is a merge conflict that needs to be fixed.

By the way you should not use your master branch for writing a pull request.

InCogNiTo124 commented:
> By the way you should not use your master branch for writing a pull request.

Thanks for the tip, I'm still learning. Will this be a serious problem?

oscarbenjamin commented:
> > By the way you should not use your master branch for writing a pull request.
> 
> Thanks for the tip, I'm still learning. Will this be a serious problem?

The most obvious problem is it makes it difficult to write two PRs simultaneously. Your master branch is now tied to this PR until it gets merged (some PRs never get merged!).

Afterwards though it means that the history on your master branch no longer corresponds to the history of the main repo. There are ways to fix this but I suggest leaving it for now until this gets merged (I think it will).

Step 1 is to fix the merge conflicts with master (i.e. the main repo master - if you didn't commit to your master then there would be no distinction).

InCogNiTo124 commented:
I can't put `#noqa` in `interactive/session.py` nor `utilities/_compilation/compilation.py` because the tests that fail overthere are annotated with `#doctest: +SKIP` so `doctest` module doing the actual testing interprets it as its option (which then makes it fail when running `./bin/doctest`).

oscarbenjamin commented:
The doctest for `compile_link_import_py_ext` is broken so flake8 is right to report an error there. I suggest just to remove the doctest there.

For session.py I suggest ignoring those errors for the file:
https://stackoverflow.com/a/54454433/9450991

InCogNiTo124 commented:
Something subtle is happening in `core/basic.py:579`, this is the only doctest to be failing

oscarbenjamin commented:
That doctest should give False but I suspect that the surrounding text would have been changed on master as well

oscarbenjamin commented:
I think that in resolving the merge conflict you've re-added part of the docstring that was removed on master:
```diff
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index 53beba2396..14b5fceb0d 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -554,7 +554,7 @@ def as_dummy(self):
         ========
 
         >>> from sympy import Integral, Symbol
-        >>> from sympy.abc import x, y
+        >>> from sympy.abc import x
         >>> r = Symbol('r', real=True)
         >>> Integral(r, (r, x)).as_dummy()
         Integral(_0, (_0, x))
@@ -570,6 +570,14 @@ def as_dummy(self):
         a property, `bound_symbols` that returns those symbols
         appearing in the object.
 
+        Lambda and Subs have bound symbols, but because of how they
+        are cached, they already compare the same regardless of their
+        bound symbols:
+
+        >>> from sympy.abc import x, y
+        >>> from sympy import Lambda
+        >>> Lambda(x, x + 1) == Lambda(y, y + 1)
+        True
         """
         from sympy.core.symbol import Dummy, Symbol
         def can(x):
```

oscarbenjamin commented:
This is how the docstring looks on current master:
https://github.com/sympy/sympy/blob/b0384246d34381a189922b19b2a82a447fc9d180/sympy/core/basic.py#L546-L573

InCogNiTo124 commented:
Is it done now?

oscarbenjamin commented:
This has merge conflicts again already. This is the difficulty with a PR that touches so many parts of the codebase.

Also one of the earliest commits deletes a file which is then readded later. Unfortunately it is not enough to add the file in a subsequent commit as this removes all of the "history" from the file. 

Compare master:
https://github.com/sympy/sympy/blame/master/bin/test_travis.sh
With the PR:
https://github.com/sympy/sympy/blame/e6cc7fa5902c8e3673f9a5b1f45332862489c3ae/bin/test_travis.sh

The commits will need to be squashed or recommitted so that the file is not removed in the history.

oscarbenjamin committed:
quality(doc): Make doctests pass flake8 checking

Mostly removes no longer needed imports.

Co-authored-by: Marijan Smetko <marijansmetko123@gmail.com>

oscarbenjamin committed:
quality(doc): Test doctests with flake8 on Travis

Mostly removes no longer needed imports.

Co-authored-by: Marijan Smetko <marijansmetko123@gmail.com>

oscarbenjamin commented on bin/test_travis.sh:
```diff
@@ -7,6 +7,7 @@ set -x
 
 if [[ "${TEST_FLAKE8}" == "true" ]]; then
     flake8 sympy;
+    flake8 --doctests --per-file-ignores='sympy/interactive/session.py:F821' sympy;
```
Does this run the normal flake8 checking twice?

InCogNiTo124 replied:
Yes, I found no way for flake to simoultaneously check for docs and code

InCogNiTo124 replied:
I'm not sure. It seems so, but I can't find anything in the `flake8` documentation that says whether or not the code is also checked with the `--doctests` flag.

edit: I found a reference here: https://bugs.launchpad.net/pyflakes/+bug/1223150 so it seems that the normal checking truly is done twice.

oscarbenjamin commented:
There were problems with the way that the commits had been squashed in combination with merging with master. I think I've resolved them and force-pushed onto this PR.

You should use these commands to synchronise the changes to your computer:
```
git fetch origin
git reset origin/master --hard
```
(Don't use `git push` as it will try to merge)

InCogNiTo124 committed:
Remove unnecessary flake8 invoking

It's hardly documented, but making flake8 check for doctests does actually check the whole file. Therefore, the previous version was checking on the files twice, and doctests only once.

oscarbenjamin commented:
Can the per-file-ignores be put in setup.cfg?

InCogNiTo124 commented:
> Can the per-file-ignores be put in setup.cfg?

[Yes](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-per-file-ignores). Should I refactor that option there?

oscarbenjamin commented:
Yes. There already others there

InCogNiTo124 committed:
Refactor flake8 flags to setup.cfg

oscarbenjamin commented:
I've just been through the whole diff. This is good to merge once tests pass (it should be merged quickly before new conflicts arise)

oscarbenjamin commented:
Thanks @InCogNiTo124! I know it's not easy to run through the whole codebase like this.

This is merged now.

InCogNiTo124 commented:
> Thanks @InCogNiTo124! I know it's not easy to run through the whole codebase like this.
> 
> This is merged now.

This was my first opensource PR and it as truly a baptism by fire, but I really learned a lot :D

oscarbenjamin commented:
You've handled this very well. I look forward to seeing more :)

Now that this is in you should fix your master branch e.g. with:
```
git reset HEAD~100 --hard
git fetch origin
git pull
```
For future PRs create a branch for each PR so you can keep your master branch clean.

InCogNiTo124 commented:
Thanks :) I'll come back for some more easy fixes

oscarbenjamin commented:
The instruction I just gave won't work actually because your master is pushed to origin (your fork). You need to add the main sympy repo as a remote (called upstream) and then do:
```
git reset HEAD~100 --hard
git fetch upstream
git merge upstream/master
```
